File: web_version/pmugg_files/selector.js
// Selects points

ms.selector = function(driver, internalSelector) {
	this.selection = null;
 	this.driver = driver;
	this.internalSelector = internalSelector;
	this.observers = [];
};

ms.selector.brush = null;

ms.selector.getBrush = function() {
	if (!ms.selector.brush) {
		ms.selector.brush = new ms.brush('#888', '#888', () => {});
	}
	return ms.selector.brush;
}

ms.selector.prototype.register = function(observer) {
	this.observers.push(observer);
};

ms.selector.prototype.remove = function(observer) {
	ms.remove(observer, this.observers);
};

ms.selector.prototype.notify = function() {
	for (var i = 0; i < this.observers.length; i++) {
		this.observers[i].redraw(this);
	}
};

ms.selector.prototype.getRenderables = function() {
	var renderables = this.driver.getRenderables().slice();
	if (this.selection) {
		renderables.push(this.selection);
	}
	return renderables;
};

ms.selector.prototype.getVertices = function () {
	return this.driver.getVertices();
};

ms.selector.prototype.getEdges = function () {
	var edges = this.driver.getEdges();
	if (this.selection) {
		edges = edges.concat(this.selection.getEdges());
	}
	return edges;
};

ms.selector.prototype.selectNone = function() {
	this.driver.selectNone();
	this.selection = null;
	this.notify();
};

ms.selector.prototype.selectAll = function() {
	this.driver.selectAll();
	this.selection = null;
	this.notify();
};

ms.selector.prototype.unfreezeAll = function() {
	this.driver.unfreezeAll();
	this.notify();
};

ms.selector.prototype.createSelectRect = function(event) {
	this.selection = new ms.shapeSet();
	for (var i = 0; i < 4; i++) {
		this.selection.vertices.push(new ms.point(event.x, event.y));
	}
	for (var i = 0; i < 4; i++) {
		var edge = new ms.edge(this.selection.vertices[i], this.selection.vertices[(i + 1) % 4]);
		edge.setBrush(ms.selector.getBrush());
		this.selection.edges.push(edge);
	}
	this.notify();
};

ms.selector.prototype.moveSelectRect = function(position) {
	if (!this.selection) {
		return;
	}
	var vertices = this.selection.vertices;
	vertices[1].position.x = position.x;
	vertices[2].position.x = position.x;
	vertices[2].position.y = position.y;
	vertices[3].position.y = position.y;
	this.notify();
};

ms.selector.prototype.createSelectLasso = function(position) {
	this.selection = new ms.shapeSet();
	var v = new ms.exampleVertex(position.x, position.y);
	this.selection.addVertex(v);
};

ms.selector.prototype.closeLasso = function(event) {
	if (!this.selection) {
		return;
	}
	var vertices = this.selection.vertices;
	var lastVertex = vertices[vertices.length - 1];
	var firstVertex = vertices[0];
	this.selection.edges.push(new ms.edge(lastVertex, firstVertex));
	this.notify();
};

ms.selector.prototype.addLasso = function(event) {
	if (!this.selection) {
		return;
	}
	this.selection.addPolyLine(event.x, event.y);
	this.notify();
};

ms.selector.prototype.select = function(add, subtract) {
	if (!this.selection) {
		return;
	}
	this.internalSelector.select(this.selection, add, subtract);
	this.selection = null;
	this.notify();
};

// Update an edge or vertex depending on if it was selected.
ms.selector.updateSelectable = function(selectable, add, subtract, insideSelection) {
	if (!add && !subtract) {
		if (insideSelection) {
			selectable.select();
		} else {
			selectable.deselect();
		}
	} else if (add && !selectable.getSelected() && insideSelection) {
		selectable.select();
	} else if (subtract && selectable.getSelected() && insideSelection) {
		selectable.deselect();
	}
};

ms.selector.prototype.findVertex = function(position, scale) {
	return this.internalSelector.findVertex(position, scale);
};

ms.selector.prototype.pointSelect = function(position, options) {
	var selectable = this.internalSelector.pointSelect(position, options);
	this.notify();
	return selectable;
};

File: cpp_version/pmugg/selector.cpp
#include "selector.h"
#include "view.h"
#include "canvas.h"
#include "context.h"

namespace ms {

Brush* Selector::brush = nullptr;

Selector::Selector(Driver* driver, InternalSelector* internalSelector)
    : driver(driver)
    , internalSelector(internalSelector)
    , lassoEnabled(false) {
    
    selection.active = false;
    selection.add = false;
    selection.subtract = false;
}

Selector::~Selector() {
    // Clean up any resources
}

void Selector::register_(View* view) {
    addObserver([view]() { view->redraw(nullptr); });
}

void Selector::setLasso(bool enabled) {
    lassoEnabled = enabled;
}

void Selector::selectAll() {
    driver->selectAll();
    notify();
}

void Selector::selectNone() {
    driver->selectNone();
    notify();
}

void Selector::cancel() {
    if (selection.active) {
        selection.active = false;
        notify();
    }
}

void Selector::notify() {
    for (const auto& observer : observers) {
        observer();
    }
}

void Selector::onMouseMove(const MouseEvent& event) {
    if (selection.active) {
        selection.current = event.position;
        notify();
    }
}

void Selector::onMouseDown(const MouseEvent& event) {
    if (event.button != MouseButton::LEFT) {
        return;
    }

    selection.start = event.position;
    selection.current = event.position;
    selection.active = true;
    selection.add = event.shiftKey;
    selection.subtract = event.altKey;

    if (!selection.add && !selection.subtract) {
        driver->selectNone();
    }

    notify();
}

void Selector::onMouseUp(const MouseEvent& event) {
    if (!selection.active || event.button != MouseButton::LEFT) {
        return;
    }

    if (selection.start.distanceTo(event.position) < 5) {
        // Point selection
        SelectOptions options;
        options.add = selection.add;
        options.subtract = selection.subtract;
        options.scale = event.scale;
        pointSelect(event.position, options);
    } else {
        // Area selection
        applySelection();
    }

    selection.active = false;
    notify();
}

void Selector::addObserver(const std::function<void()>& observer) {
    observers.push_back(observer);
}

void Selector::removeObserver(const std::function<void()>& observer) {
    // Remove observer by comparing function pointers
    observers.erase(
        std::remove_if(observers.begin(), observers.end(),
            [&](const auto& obs) { 
                return obs.target<void()>() == observer.target<void()>(); 
            }),
        observers.end()
    );
}

void Selector::drawSelection(const Vec2& start, const Vec2& end) {
    if (!brush) {
        return;
    }

    auto* context = Canvas::getContext();
    context->save();
    
    if (lassoEnabled) {
        // Draw lasso selection
        context->beginPath();
        context->moveTo(start.x, start.y);
        context->lineTo(end.x, end.y);
        context->setStrokeStyle(brush->getColor());
        context->setLineWidth(2);
        context->stroke();
    } else {
        // Draw rectangle selection
        float x = std::min(start.x, end.x);
        float y = std::min(start.y, end.y);
        float width = std::abs(end.x - start.x);
        float height = std::abs(end.y - start.y);

        context->setStrokeStyle(brush->getColor());
        context->setLineWidth(2);
        context->strokeRect(x, y, width, height);

        context->setFillStyle(brush->getColor());
        context->setGlobalAlpha(0.1);
        context->fillRect(x, y, width, height);
    }

    context->restore();
}

void Selector::clearSelection() {
    driver->selectNone();
    notify();
}

void Selector::applySelection() {
    if (lassoEnabled) {
        // Apply lasso selection
        std::vector<Vec2> points = {selection.start, selection.current};
        driver->selectLasso(points, selection.add, selection.subtract);
    } else {
        // Apply rectangle selection
        float x = std::min(selection.start.x, selection.current.x);
        float y = std::min(selection.start.y, selection.current.y);
        float width = std::abs(selection.current.x - selection.start.x);
        float height = std::abs(selection.current.y - selection.start.y);

        Rect selectionRect(x, y, width, height);
        driver->selectRect(selectionRect, selection.add, selection.subtract);
    }
}

Vertex* Selector::findVertex(const Vec2& position, float scale) {
    return internalSelector->findVertex(position, scale);
}

Selectable* Selector::pointSelect(const Vec2& position, const SelectOptions& options) {
    auto* selectable = internalSelector->pointSelect(position, options);
    notify();
    return selectable;
}

} // namespace ms 